pkg/metrics/providers/graphite_test.go

```diff
@@ -17,18 +17,66 @@ limitations under the License.
 package providers
 
 import (
+	"context"
 	"fmt"
 	"net/http"
 	"net/http/httptest"
+	"strings"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
 
 	flaggerv1 "github.com/fluxcd/flagger/pkg/apis/flagger/v1beta1"
+	fakeFlagger "github.com/fluxcd/flagger/pkg/client/clientset/versioned/fake"
 )
 
+func graphiteFake(query string) fakeClients {
+
+	provider := flaggerv1.MetricTemplateProvider{
+		Type:      "graphite",
+		Address:   "http://graphite:8080",
+		SecretRef: &corev1.LocalObjectReference{Name: "graphite"},
+	}
+
+	template := &flaggerv1.MetricTemplate{
+		TypeMeta: metav1.TypeMeta{APIVersion: flaggerv1.SchemeGroupVersion.String()},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "graphite",
+		},
+		Spec: flaggerv1.MetricTemplateSpec{
+			Provider: provider,
+			Query:    query,
+		},
+	}
+
+	flaggerClient := fakeFlagger.NewSimpleClientset(template)
+
+	secret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{APIVersion: corev1.SchemeGroupVersion.String()},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "graphite",
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: map[string][]byte{
+			"username": []byte("username"),
+			"password": []byte("password"),
+		},
+	}
+
+	kubeClient := fake.NewSimpleClientset(secret)
+
+	return fakeClients{
+		kubeClient:    kubeClient,
+		flaggerClient: flaggerClient,
+	}
+}
+
```

Comment:
☝️ Pardon my naivete, but is this all necessary? Wouldn't something like the following suffice?

```golang
graphiteProvider, err := NewGraphiteProvider(flaggerv1.MetricTemplateProvider{
  Address:   testServer.URL,
  Type:      "graphite",
  SecretRef: &corev1.LocalObjectReference{Name: "graphite"},
}, map[string][]byte{
    "username": []byte("a-username"),
    "password": []byte("a-password"),
  }
)
```

...or maybe even something like this:

```golang
graphiteProvider, err := NewGraphiteProvider(flaggerv1.MetricTemplateProvider{
  Address:   testServer.URL,
}, map[string][]byte{})
```

Reply:
Likely. I just copied what is done in the prometheus test. 

Looking at the code, I think the goal was to more closely simulate what flagger actually does (extracting data from various k8s resources).